test/test_ext_logins.js

```diff
@@ -0,0 +1,166 @@
+"use strict";
+
+Cu.import("resource://gre/modules/AddonManager.jsm");
+Cu.import("resource://gre/modules/FileUtils.jsm");
+Cu.import("resource://gre/modules/osfile.jsm");
+Cu.import("resource://gre/modules/Services.jsm");
+
+function makeLoginInfo(data) {
+  let info = Cc["@mozilla.org/login-manager/loginInfo;1"].createInstance(Ci.nsILoginInfo);
+  for (let field of Object.keys(data)) {
+    info[field] = data[field];
+  }
+  return info;
+}
+
+// record is a simple js object from the extension api, info is an nsILoginInfo
+// accounts for field name differences
+function checkRecord(record, info) {
+  equal(record.formSubmitURL, info.formSubmitURL);
+  equal(record.origin, info.hostname);
+  equal(record.realm, info.httpRealm);
+  equal(record.username, info.username);
+  equal(record.password, info.password);
+  equal(record.usernameField, info.usernameField);
+  equal(record.passwordField, info.passwordField);
+}
+
+function loadApiExtension() {
+  notEqual(_TEST_FILE, undefined, "_TEST_FILE is set");
+  let testFile = new FileUtils.File(_TEST_FILE);
+  equal(testFile.isSymlink(), true, "_TEST_FILE is a symlink");
+  testFile = new FileUtils.File(testFile.target);
+  // testFile.target should be this file, so its parent is the test
+  // directory and parent.parent is the top level for the api extension
+  let apiExtensionDir = testFile.parent.parent;
+  do_print(`mapped test file ${_TEST_FILE} to api extension directory ${apiExtensionDir.path}\n`);
+
+  return AddonManager.installTemporaryAddon(apiExtensionDir);
+}
+
+add_task(function* test_logins() {
+  yield ExtensionTestUtils.startAddonManager();
+
+  let apiExtension = yield loadApiExtension();
+
+  function background() {
+    browser.test.onMessage.addListener(function(msg, args) {
+      let match = msg.match(/^(\w+)\.request$/);
+      if (!match) {
+        return;
+      }
+      let cmd = match[1];
+      Promise.resolve().then(() => browser.logins[cmd](...args))
+        .then(results => {
+          browser.test.sendMessage(`${cmd}.done`, {results});
+        }, err => {
+          browser.test.sendMessage(`${cmd}.done`, {errmsg: err.message});
+        });
+    });
+    browser.test.sendMessage("ready");
+  }
+
+  function run(ext, cmd, ...args) {
+    let promise = ext.awaitMessage(`${cmd}.done`);
+    ext.sendMessage(`${cmd}.request`, args);
+    return promise;
+  }
+
+  let privilegedExtension = ExtensionTestUtils.loadExtension({
+    background,
+    manifest: {
+      permissions: ["experiments.logins", "logins", "<all_urls>"],
+    },
+  });
```

Comment:
Yes, it uses the host permissions that are already part of an extension.  The difference between `privilegedExtension` and `unprivilegedExtension` is that the former includes the `<all_urls>` permission which gives it host permissions for everything and the latter does not.  My reasoning was that we should test the basic mechanics of host permission (there are such tests below) but that fine-grained testing of different types of host permissions should be in a separate unit test of permissions or MatchPattern.  However, with a quick glance, I can't find such a test :(